File: backend/routes/contabilidad.py
"""routes/contabilidad.py — Módulo contable (#809).

Transporte HTTP fino sobre el motor `backend/contabilidad/` (igual que
`routes/reportes.py` sobre `reportes/`). Fase 1: cuentas (CRUD) + saldos + tablero
mínimo. Acceso = cualquier admin (`require_admin`); el actor de auditoría sale de
`admin.get("email")`.
"""

import logging

from fastapi import APIRouter, Request, HTTPException, UploadFile, File
from pydantic import BaseModel

from database import get_db
from admin_guard import require_admin
from contabilidad.cuentas import (
    crear_cuenta,
    desactivar_cuenta,
    editar_cuenta,
    listar_cuentas,
)
from contabilidad.saldos import saldos as _saldos
from contabilidad.categorias import crear_categoria, listar_categorias
from contabilidad.movimientos import (
    anular_movimiento,
    beneficiarios_usados,
    cobros_mensuales,
    crear_movimiento,
    editar_movimiento,
    gastos_por_categoria,
    listar_movimientos,
    obtener_movimiento,
)
from contabilidad.tablero import tablero as _tablero
from contabilidad.pyl import ganancia_neta
from contabilidad.rendicion import rendicion as _rendicion, saldar as _saldar
from contabilidad.cierres import cerrar_mes as _cerrar_mes, reabrir_mes as _reabrir_mes
from contabilidad.reconciliacion import reconciliar as _reconciliar

logger = logging.getLogger(__name__)
router = APIRouter()


class CuentaCreate(BaseModel):
    nombre: str
    tipo: str
    socio: str | None = None
    moneda: str = "ARS"
    saldo_inicial: int = 0
    fecha_apertura: str | None = None
    orden: int = 0


class CuentaUpdate(BaseModel):
    nombre: str | None = None
    saldo_inicial: int | None = None
    fecha_apertura: str | None = None
    orden: int | None = None
    activa: bool | None = None


@router.get("/admin/contabilidad/saldos")
def get_saldos(request: Request):
    """Saldos de todas las cuentas activas + total disponible. Los ingresos por
    alquiler ya vienen derivados de `alquiler_pagos`."""
    require_admin(request)
    with get_db() as conn:
        return _saldos(conn)


@router.get("/admin/contabilidad/tablero")
def get_tablero(request: Request, mes: str | None = None):
    """Tablero financiero: plata disponible por caja + ganancia neta del mes +
    rendición pendiente. `mes` opcional ('YYYY-MM', default: mes actual)."""
    require_admin(request)
    with get_db() as conn:
        try:
            return _tablero(conn, mes)
        except ValueError as e:
            raise HTTPException(400, str(e))


@router.get("/admin/contabilidad/cuentas")
def get_cuentas(request: Request, incluir_inactivas: bool = False):
    """Cuentas crudas (para administrar). `incluir_inactivas=true` trae también las
    dadas de baja."""
    require_admin(request)
    with get_db() as conn:
        return {"cuentas": listar_cuentas(conn, incluir_inactivas=incluir_inactivas)}


@router.post("/admin/contabilidad/cuentas")
def post_cuenta(request: Request, body: CuentaCreate):
    """Crea una cuenta/caja nueva."""
    admin = require_admin(request)
    with get_db() as conn:
        try:
            cuenta = crear_cuenta(
                conn,
                nombre=body.nombre,
                tipo=body.tipo,
                socio=body.socio,
                moneda=body.moneda,
                saldo_inicial=body.saldo_inicial,
                fecha_apertura=body.fecha_apertura,
                orden=body.orden,
                por=admin.get("email"),
            )
            conn.commit()
            return cuenta
        except ValueError as e:
            conn.rollback()
            raise HTTPException(400, str(e))
        except Exception:
            conn.rollback()
            raise


@router.patch("/admin/contabilidad/cuentas/{cuenta_id}")
def patch_cuenta(request: Request, cuenta_id: int, body: CuentaUpdate):
    """Edita una cuenta. No se puede cambiar tipo ni socio."""
    admin = require_admin(request)
    campos = {k: v for k, v in body.model_dump(exclude_unset=True).items() if v is not None
              or k == "activa"}
    with get_db() as conn:
        try:
            cuenta = editar_cuenta(conn, cuenta_id, campos=campos, por=admin.get("email"))
            conn.commit()
            return cuenta
        except ValueError as e:
            conn.rollback()
            raise HTTPException(400, str(e))
        except Exception:
            conn.rollback()
            raise


@router.delete("/admin/contabilidad/cuentas/{cuenta_id}")
def delete_cuenta(request: Request, cuenta_id: int):
    """Baja lógica de una cuenta (falla si tiene saldo distinto de cero)."""
    admin = require_admin(request)
    with get_db() as conn:
        try:
            cuenta = desactivar_cuenta(conn, cuenta_id, por=admin.get("email"))
            conn.commit()
            return cuenta
        except ValueError as e:
            conn.rollback()
            raise HTTPException(400, str(e))
        except Exception:
            conn.rollback()
            raise


# ── Categorías de gasto ─────────────────────────────────────────────────────

class CategoriaCreate(BaseModel):
    nombre: str


@router.get("/admin/contabilidad/categorias")
def get_categorias(request: Request):
    require_admin(request)
    with get_db() as conn:
        return {"categorias": listar_categorias(conn)}


@router.get("/admin/contabilidad/beneficiarios")
def get_beneficiarios(request: Request):
    """Beneficiarios ya usados, para el autocompletado del formulario."""
    require_admin(request)
    with get_db() as conn:
        return {"beneficiarios": beneficiarios_usados(conn)}


@router.post("/admin/contabilidad/categorias")
def post_categoria(request: Request, body: CategoriaCreate):
    require_admin(request)
    with get_db() as conn:
        try:
            cat = crear_categoria(conn, body.nombre)
            conn.commit()
            return cat
        except ValueError as e:
            conn.rollback()
            raise HTTPException(400, str(e))
        except Exception:
            conn.rollback()
            raise


# ── Movimientos (gasto / transferencia / retiro / aporte / ajuste) ──────────

class MovimientoCreate(BaseModel):
    tipo: str
    monto: int
    cuenta_origen_id: int | None = None
    cuenta_destino_id: int | None = None
    categoria_id: int | None = None
    metodo: str | None = None
    fecha: str | None = None
    nota: str | None = None
    beneficiario: str | None = None


class MovimientoUpdate(BaseModel):
    monto: int | None = None
    cuenta_origen_id: int | None = None
    cuenta_destino_id: int | None = None
    categoria_id: int | None = None
    metodo: str | None = None
    fecha: str | None = None
    nota: str | None = None
    beneficiario: str | None = None


class AnularBody(BaseModel):
    motivo: str


@router.get("/admin/contabilidad/movimientos")
def get_movimientos(
    request: Request,
    tipo: str | None = None,
    cuenta_id: int | None = None,
    categoria_id: int | None = None,
    beneficiario: str | None = None,
    desde: str | None = None,
    hasta: str | None = None,
    incluir_anulados: bool = False,
    limit: int = 500,
):
    require_admin(request)
    with get_db() as conn:
        # Movimientos manuales (salvo que se filtre solo cobros).
        movs = (
            []
            if tipo == "cobro"
            else listar_movimientos(
                conn, tipo=tipo, cuenta_id=cuenta_id, categoria_id=categoria_id,
                beneficiario=beneficiario, desde=desde, hasta=hasta,
                incluir_anulados=incluir_anulados, limit=limit,
            )
        )
        # Cobros de pedidos agregados por mes (read-only) — se muestran junto a los
        # movimientos salvo que se filtre por tipo manual / categoría / beneficiario
        # (los cobros no tienen esos atributos).
        cobros: list = []
        if tipo in (None, "", "cobro") and not categoria_id and not beneficiario:
            cobrador = None
            if cuenta_id:
                row = conn.execute("SELECT socio FROM cuentas WHERE id = %s", (cuenta_id,)).fetchone()
                cobrador = row["socio"] if row else None
                # Caja sin cobrador (Efectivo/Banco/Dólares) → no recibe cobros.
                if not cobrador:
                    cobros = []
                    return {"movimientos": movs, "cobros": cobros, "count": len(movs)}
            cobros = cobros_mensuales(conn, desde=desde, hasta=hasta, cobrador=cobrador)
    return {"movimientos": movs, "cobros": cobros, "count": len(movs)}


@router.post("/admin/contabilidad/movimientos")
def post_movimiento(request: Request, body: MovimientoCreate):
    """Registra un movimiento (gasto/transferencia/retiro/aporte/ajuste)."""
    admin = require_admin(request)
    with get_db() as conn:
        try:
            mov = crear_movimiento(
                conn, tipo=body.tipo, monto=body.monto,
                cuenta_origen_id=body.cuenta_origen_id,
                cuenta_destino_id=body.cuenta_destino_id,
                categoria_id=body.categoria_id, metodo=body.metodo,
                fecha=body.fecha, nota=body.nota, beneficiario=body.beneficiario,
                por=admin.get("email"),
            )
            conn.commit()
            return mov
        except ValueError as e:
            conn.rollback()
            raise HTTPException(400, str(e))
        except Exception:
            conn.rollback()
            raise


@router.patch("/admin/contabilidad/movimientos/{mov_id}")
def patch_movimiento(request: Request, mov_id: int, body: MovimientoUpdate):
    admin = require_admin(request)
    campos = body.model_dump(exclude_unset=True)
    with get_db() as conn:
        try:
            mov = editar_movimiento(conn, mov_id, campos=campos, por=admin.get("email"))
            conn.commit()
            return mov
        except ValueError as e:
            conn.rollback()
            raise HTTPException(400, str(e))
        except Exception:
            conn.rollback()
            raise


@router.post("/admin/contabilidad/movimientos/{mov_id}/anular")
def anular_mov(request: Request, mov_id: int, body: AnularBody):
    """Soft-delete con motivo: la plata nunca se borra, queda anulada y trazable."""
    admin = require_admin(request)
    with get_db() as conn:
        try:
            mov = anular_movimiento(conn, mov_id, motivo=body.motivo, por=admin.get("email"))
            conn.commit()
            return mov
        except ValueError as e:
            conn.rollback()
            raise HTTPException(400, str(e))
        except Exception:
            conn.rollback()
            raise


_COMPROBANTE_MAX = 5 * 1024 * 1024  # 5 MB
_COMPROBANTE_TIPOS = ("application/pdf", "image/jpeg", "image/png", "image/webp")


@router.post("/admin/contabilidad/movimientos/{mov_id}/comprobante")
async def subir_comprobante(request: Request, mov_id: int, file: UploadFile = File(...)):
    """Adjunta un comprobante (PDF o imagen) a un movimiento. Sube a R2 reusando
    la infra de media (`services/media/storage`)."""
    require_admin(request)
    content = await file.read()
    if len(content) > _COMPROBANTE_MAX:
        raise HTTPException(400, "El comprobante supera los 5 MB.")
    ctype = (file.content_type or "").lower()
    if ctype not in _COMPROBANTE_TIPOS:
        raise HTTPException(400, "Formato no soportado (subí un PDF o una imagen).")

    with get_db() as conn:
        mov = obtener_movimiento(conn, mov_id)
        if not mov:
            raise HTTPException(404, "El movimiento no existe.")
        try:
            from services.media.service import store_raw_document
            from services.media import storage

            # Borrado best-effort del comprobante anterior.
            vieja = mov.get("comprobante_key")
            if vieja:
                storage.delete_object(vieja, private=True)

            key, presigned = store_raw_document(
                content,
                kind="comprobante-contabilidad",
                ref=str(mov_id),
                content_type=ctype,
                presign_expires=3600,
            )
            conn.execute(
                "UPDATE movimientos SET comprobante_url = NULL, comprobante_key = %s, "
                "updated_at = CURRENT_TIMESTAMP WHERE id = %s",
                (key, mov_id),
            )
            conn.commit()
            return {"comprobante_url": presigned, "comprobante_key": key}
        except HTTPException:
            conn.rollback()
            raise
        except Exception:
            conn.rollback()
            logger.exception("No se pudo subir el comprobante")
            raise HTTPException(502, "No se pudo subir el comprobante")


@router.get("/admin/contabilidad/gastos")
def get_gastos(request: Request, desde: str | None = None, hasta: str | None = None):
    """Gastos agrupados por categoría en la ventana (para el tablero / análisis)."""
    require_admin(request)
    with get_db() as conn:
        filas = gastos_por_categoria(conn, desde, hasta)
    return {"por_categoria": filas, "total": sum(f["monto"] for f in filas)}


@router.get("/admin/contabilidad/pyl/{mes}")
def get_pyl(request: Request, mes: str):
    """Ganancia neta del mes (ingresos devengados − gastos)."""
    require_admin(request)
    with get_db() as conn:
        try:
            return ganancia_neta(conn, mes)
        except ValueError as e:
            raise HTTPException(400, str(e))


@router.get("/admin/contabilidad/reporte/{mes}")
def get_reporte_mensual(request: Request, mes: str):
    """Reporte mensual completo de Rambla: devengado + percibido + gastos +
    ganancia + cargos de socios + cuenta corriente, derivado del motor."""
    require_admin(request)
    from contabilidad.reporte_mensual import reporte_mensual

    with get_db() as conn:
        try:
            return reporte_mensual(conn, mes)
        except ValueError as e:
            raise HTTPException(400, str(e))


# ── Rendición de cuentas mensual entre socios ───────────────────────────────

class SaldarBody(BaseModel):
    de: str
    a: str
    monto: int
    metodo: str | None = None
    fecha: str | None = None
    nota: str | None = None


@router.get("/admin/contabilidad/rendicion/{mes}")
def get_rendicion(request: Request, mes: str):
    """Rendición del mes: cuánto le corresponde y cuánto cobró cada parte, el
    saldo pendiente, los movimientos sugeridos y lo ya rendido."""
    require_admin(request)
    with get_db() as conn:
        try:
            return _rendicion(conn, mes)
        except ValueError as e:
            raise HTTPException(400, str(e))


@router.post("/admin/contabilidad/rendicion/{mes}/saldar")
def post_saldar(request: Request, mes: str, body: SaldarBody):
    """Registra un saldado real de rendición (transferencia entre las cajas de las
    partes, marcada como rendición en el libro)."""
    admin = require_admin(request)
    with get_db() as conn:
        try:
            mov = _saldar(
                conn, mes, de=body.de, a=body.a, monto=body.monto,
                metodo=body.metodo, fecha=body.fecha, nota=body.nota,
                por=admin.get("email"),
            )
            conn.commit()
            return mov
        except ValueError as e:
            conn.rollback()
            raise HTTPException(400, str(e))
        except Exception:
            conn.rollback()
            raise


# ── Reconciliación + cierre contable del mes ────────────────────────────────

@router.get("/admin/contabilidad/reconciliacion")
def get_reconciliacion(request: Request):
    """Semáforo de confianza: chequea que la plata del módulo cuadre."""
    require_admin(request)
    with get_db() as conn:
        return _reconciliar(conn)


@router.post("/admin/contabilidad/cierres/{mes}")
def post_cierre(request: Request, mes: str):
    """Cierra un mes contable: congela la foto y traba la edición de sus
    movimientos. Idempotente."""
    admin = require_admin(request)
    with get_db() as conn:
        try:
            return _cerrar_mes(conn, mes, admin.get("email"))
        except ValueError as e:
            conn.rollback()
            raise HTTPException(400, str(e))


@router.delete("/admin/contabilidad/cierres/{mes}")
def delete_cierre(request: Request, mes: str):
    """Reabre un mes contable cerrado."""
    require_admin(request)
    with get_db() as conn:
        try:
            reabierto = _reabrir_mes(conn, mes)
        except ValueError as e:
            raise HTTPException(400, str(e))
    return {"mes": mes, "cerrado": False, "reabierto": reabierto}

File: frontend/src/lib/admin/api/contabilidad.ts
import { authedFetch, authedJson, authedPostJson } from "@/lib/authedFetch";
import type {
  TableroData,
  SaldosData,
  Cuenta,
  CuentaInput,
  Movimiento,
  MovimientoInput,
  CobroMensual,
  GastoCategoria,
  GastosPorCategoria,
  ReporteMensual,
  RendicionData,
  ReconciliacionContable,
} from "./types";

export const contabilidadMethods = {
  // Contabilidad (#809) — Fase 1: cuentas/cajas con saldo + tablero. Los ingresos
  // por alquiler salen derivados de alquiler_pagos (no se cargan a mano).
  getTablero: (mes?: string) =>
    authedJson<TableroData>(`/api/admin/contabilidad/tablero${mes ? `?mes=${mes}` : ""}`),
  getSaldos: () => authedJson<SaldosData>("/api/admin/contabilidad/saldos"),
  listCuentas: (incluirInactivas = false) =>
    authedJson<{ cuentas: Cuenta[] }>(
      `/api/admin/contabilidad/cuentas${incluirInactivas ? "?incluir_inactivas=true" : ""}`,
    ),
  createCuenta: (data: CuentaInput) =>
    authedPostJson<Cuenta>("/api/admin/contabilidad/cuentas", data),
  updateCuenta: (id: number, data: Partial<CuentaInput> & { activa?: boolean }) =>
    authedJson<Cuenta>(`/api/admin/contabilidad/cuentas/${id}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(data),
    }),
  deactivateCuenta: async (id: number) => {
    const res = await authedFetch(`/api/admin/contabilidad/cuentas/${id}`, { method: "DELETE" });
    if (!res.ok && res.status !== 204) {
      const detail = await res.json().catch(() => ({}));
      throw new Error(detail?.detail ?? `DELETE → ${res.status}`);
    }
    return res.json().catch(() => ({}));
  },

  // Contabilidad — movimientos (gasto/transferencia/retiro/aporte/ajuste) + categorías
  listMovimientos: (params?: {
    tipo?: string;
    cuenta_id?: number;
    categoria_id?: number;
    beneficiario?: string;
    desde?: string;
    hasta?: string;
    incluir_anulados?: boolean;
  }) => {
    const sp = new URLSearchParams();
    if (params?.tipo) sp.set("tipo", params.tipo);
    if (params?.cuenta_id) sp.set("cuenta_id", String(params.cuenta_id));
    if (params?.categoria_id) sp.set("categoria_id", String(params.categoria_id));
    if (params?.beneficiario) sp.set("beneficiario", params.beneficiario);
    if (params?.desde) sp.set("desde", params.desde);
    if (params?.hasta) sp.set("hasta", params.hasta);
    if (params?.incluir_anulados) sp.set("incluir_anulados", "true");
    const qs = sp.toString();
    return authedJson<{ movimientos: Movimiento[]; cobros: CobroMensual[]; count: number }>(
      `/api/admin/contabilidad/movimientos${qs ? `?${qs}` : ""}`,
    );
  },
  listBeneficiarios: () =>
    authedJson<{ beneficiarios: string[] }>("/api/admin/contabilidad/beneficiarios"),
  createMovimiento: (data: MovimientoInput) =>
    authedPostJson<Movimiento>("/api/admin/contabilidad/movimientos", data),
  updateMovimiento: (id: number, data: Partial<MovimientoInput>) =>
    authedJson<Movimiento>(`/api/admin/contabilidad/movimientos/${id}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(data),
    }),
  anularMovimiento: (id: number, motivo: string) =>
    authedPostJson<Movimiento>(`/api/admin/contabilidad/movimientos/${id}/anular`, { motivo }),
  uploadComprobante: async (id: number, file: File): Promise<{ comprobante_url: string }> => {
    const fd = new FormData();
    fd.append("file", file);
    const res = await authedFetch(`/api/admin/contabilidad/movimientos/${id}/comprobante`, {
      method: "POST",
      body: fd,
    });
    const json = await res.json().catch(() => ({}));
    if (!res.ok) throw new Error(json?.detail ?? `POST → ${res.status}`);
    return json;
  },
  listGastoCategorias: () =>
    authedJson<{ categorias: GastoCategoria[] }>("/api/admin/contabilidad/categorias"),
  createGastoCategoria: (nombre: string) =>
    authedPostJson<GastoCategoria>("/api/admin/contabilidad/categorias", { nombre }),
  getGastos: (desde?: string, hasta?: string) => {
    const sp = new URLSearchParams();
    if (desde) sp.set("desde", desde);
    if (hasta) sp.set("hasta", hasta);
    const qs = sp.toString();
    return authedJson<GastosPorCategoria>(`/api/admin/contabilidad/gastos${qs ? `?${qs}` : ""}`);
  },
  getReporteMensual: (mes: string) =>
    authedJson<ReporteMensual>(`/api/admin/contabilidad/reporte/${mes}`),
  // Rendición de cuentas mensual entre socios
  getRendicion: (mes: string) =>
    authedJson<RendicionData>(`/api/admin/contabilidad/rendicion/${mes}`),
  saldarRendicion: (
    mes: string,
    body: { de: string; a: string; monto: number; metodo?: string | null; nota?: string | null },
  ) => authedPostJson<Movimiento>(`/api/admin/contabilidad/rendicion/${mes}/saldar`, body),
  getPyl: (mes: string) =>
    authedJson<{
      mes: string;
      ingresos: number;
      gastos: number;
      ganancia_neta: number;
      gastos_por_categoria: { categoria: string; monto: number }[];
    }>(`/api/admin/contabilidad/pyl/${mes}`),
  getReconciliacionContable: () =>
    authedJson<ReconciliacionContable>("/api/admin/contabilidad/reconciliacion"),
  cerrarMesContable: (mes: string) =>
    authedJson<{ mes: string; cerrado: boolean }>(`/api/admin/contabilidad/cierres/${mes}`, {
      method: "POST",
    }),
  reabrirMesContable: async (mes: string) => {
    const res = await authedFetch(`/api/admin/contabilidad/cierres/${mes}`, { method: "DELETE" });
    if (!res.ok && res.status !== 204) {
      const detail = await res.json().catch(() => ({}));
      throw new Error(detail?.detail ?? `DELETE → ${res.status}`);
    }
    return res.json().catch(() => ({}));
  },
};
